std/algorithm/sorting.d

```diff
@@ -2994,6 +2994,165 @@ unittest
     }
 }
 
+// medianOf
+/**
+Computes the median of 2 to 5 arbitrary indexes in random-access range `r`
+using hand-written specialized algorithms. The indexes must be distinct (if not,
+behavior is implementation-defined). The function also partitions the elements
+involved around the median, e.g. $(D median(r, a, b, c)) not only fills `r[b]`
+with the median of `r[a]`, `r[b]`, and `r[c]`, but also puts the minimum in
+`r[a]` and the maximum in `r[c]`.
+
+Params:
+less = The comparison predicate used, modeled as a $(LUCKY strict weak
+ordering) (irreflexive, antisymmetric, transitive, and implying a transitive
+equivalence).
+flag = Used only for even values of `T.length`. If `No.leanRight`, the median
+"leans left", meaning $(D median(r, a, b, c, d)) puts the lower median of the
+four in `r[b]`, the minimum in `r[a]`, and the two others in `r[c]` and `r[d]`.
+Conversely, $(D median!("a < b", Yes.leanRight)(r, a, b, c, d)) puts the upper
+median of the four in `r[c]`, the maximum in `r[d]`, and the two others in
+`r[a]` and `r[b]`.
+r = The range containing the indexes.
+i = Two to five indexes inside `r`.
+*/
+void medianOf(
+        alias less = "a < b",
+        Flag!"leanRight" flag = No.leanRight,
+        Range,
+        Index...)
+    (Range r, Index i)
+if (isRandomAccessRange!Range && Index.length >= 2 && Index.length <= 5 &&
+    allSatisfy!(isIntegral, Index))
+{
+    assert(r.length >= Index.length);
+    import std.functional : binaryFun;
+    alias lt = binaryFun!less;
+    enum k = Index.length;
+    import std.algorithm : swapAt;
+
+    alias a = i[0];
+    static if (k >= 2)
+    {
+        alias b = i[1];
+        assert(a != b);
+    }
+    static if (k >= 3)
+    {
+        alias c = i[2];
+        assert(a != c && b != c);
+    }
+    static if (k >= 4)
+    {
+        alias d = i[3];
+        assert(a != d && b != d && c != d);
+    }
+    static if (k >= 5)
+    {
+        alias e = i[4];
+        assert(a != e && b != e && c != e && d != e);
```

Comment:
> The indexes must be distinct

Does it also make sense for them to be unordered w.r.t. `less`? If no, then you could save some lines by using [`strictlyOrdered`](http://dlang.org/phobos-prerelease/std_algorithm_sorting#.strictlyOrdered):
```
assert(strictlyOrdered!less(i));
alias a = i[0];
static if (k >= 2)
    alias b = i[1];
static if (k >= 3)
    alias c = i[2];
static if (k >= 4)
    alias d = i[3];
static if (k == 5)
    alias e = i[4];
```